From b272d38e3725afdaa153395693114ed6cd3136e1 Mon Sep 17 00:00:00 2001
From: Vishnu Nair <vishnun@google.com>
Date: Wed, 2 Feb 2022 17:13:21 +0000
Subject: [PATCH] Drop input for toast and child surfaces

Toasts that do not have the trustedOverlay flag should not receive input.
These windows should not have any children, so force this hierarchy of
windows to drop all input by setting a flag on the toast window state
which will apply the DROP_INPUT flag on all windows with an input
channel. This is to prevent malicious apps from parenting surfaces with
input channels to the toast window.

Test: show toast and check if input feature flag DROP_INPUT id set via dumpsys
Bug: b/197296414

Change-Id: I316b76b685ca5030fd8aa91283555efcce4d6994
Merged-In: I316b76b685ca5030fd8aa91283555efcce4d6994
(cherry picked from commit 55c1473bf2fedaacf7bb8ac068e6f9b1a625b5e0)
Merged-In: I316b76b685ca5030fd8aa91283555efcce4d6994
---
 .../java/com/android/server/wm/DisplayPolicy.java | 15 +++++++++++++++
 .../android/server/wm/WindowManagerService.java   |  1 +
 2 files changed, 16 insertions(+)

diff --git a/services/core/java/com/android/server/wm/DisplayPolicy.java b/services/core/java/com/android/server/wm/DisplayPolicy.java
index 05875d971eb5..8a01d90bca29 100644
--- a/services/core/java/com/android/server/wm/DisplayPolicy.java
+++ b/services/core/java/com/android/server/wm/DisplayPolicy.java
@@ -138,6 +138,7 @@ import android.graphics.Insets;
 import android.graphics.PixelFormat;
 import android.graphics.Rect;
 import android.graphics.Region;
+import android.gui.DropInputMode;
 import android.hardware.input.InputManager;
 import android.hardware.power.V1_0.PowerHint;
 import android.os.Handler;
@@ -918,6 +919,20 @@ public class DisplayPolicy {
         }
     }
 
+    /**
+     * Add additional policy if needed to ensure the window or its children should not receive any
+     * input.
+     */
+    public void setDropInputModePolicy(WindowState win, LayoutParams attrs) {
+        if (attrs.type == TYPE_TOAST
+                && (attrs.privateFlags & PRIVATE_FLAG_TRUSTED_OVERLAY) == 0) {
+            // Toasts should not receive input. These windows should not have any children, so
+            // force this hierarchy of windows to drop all input.
+            mService.mTransactionFactory.get()
+                    .setDropInputMode(win.getSurfaceControl(), DropInputMode.ALL).apply();
+        }
+    }
+
     /**
      * @return {@code true} if the calling activity initiate toast and is visible with
      * {@link WindowManager.LayoutParams#FLAG_SHOW_WHEN_LOCKED} flag.
diff --git a/services/core/java/com/android/server/wm/WindowManagerService.java b/services/core/java/com/android/server/wm/WindowManagerService.java
index aa598d0cfdc6..9cf17e4ece5c 100644
--- a/services/core/java/com/android/server/wm/WindowManagerService.java
+++ b/services/core/java/com/android/server/wm/WindowManagerService.java
@@ -1650,6 +1650,7 @@ public class WindowManagerService extends IWindowManager.Stub
 
             win.mToken.addWindow(win);
             displayPolicy.addWindowLw(win, attrs);
+            displayPolicy.setDropInputModePolicy(win, win.mAttrs);
             if (type == TYPE_INPUT_METHOD) {
                 displayContent.setInputMethodWindowLocked(win);
                 imMayMove = false;
-- 
2.38.1.273.g43a17bfeac-goog

